RedditSharp/WebAgent.cs

```diff
@@ -112,13 +63,39 @@ public int RequestsThisBurst
         public WebAgent( string accessToken ) {
             RootDomain = OAuthDomainUrl;
             AccessToken = accessToken;
+            RateLimitUsed = 0;
+            RateLimitReset = 60;
+            RateLimitRemaining = IsOAuth() ? 60 : 30;
         }
 
         /// <inheritdoc />
-        public virtual async Task<JToken> ExecuteRequestAsync(HttpRequestMessage request)
+        public virtual async Task<JToken> ExecuteRequestAsync(Func<HttpRequestMessage> request)
         {
-            await EnforceRateLimit().ConfigureAwait(false);
-            var response = await _httpClient.SendAsync(request).ConfigureAwait(false);
+            if (request == null)
+              throw new ArgumentNullException(nameof(request));
+            const int maxTries = 20;
+            HttpResponseMessage response;
+            var tries = 0;
+            do {
+              if (RateLimitRemaining <= 0) {
+                await Task.Delay(RateLimitReset * 1000);
+              }
+              response = await _httpClient.SendAsync(request()).ConfigureAwait(false);
+              await rateLimitLock.WaitAsync().ConfigureAwait(false);
```

Comment:
I agree that this could lead to potential situations where multiple requests are fired off and come back in an incorrect order, leading to incorrect rate limit information. However, moving the lock to include the HTTP request would prevent potential parrallelization of requests, as one request would block the others.